api/server/httputils/httputils.go

```diff
@@ -49,17 +49,48 @@ func CheckForJSON(r *http.Request) error {
 	ct := r.Header.Get("Content-Type")
 
 	// No Content-Type header is ok as long as there's no Body
-	if ct == "" {
-		if r.Body == nil || r.ContentLength == 0 {
-			return nil
-		}
+	if ct == "" && (r.Body == nil || r.ContentLength == 0) {
+		return nil
 	}
 
 	// Otherwise it better be json
-	if matchesContentType(ct, "application/json") {
+	return matchesContentType(ct, "application/json")
+}
+
+// ReadJSON validates the request to have the correct content-type, and decodes
+// the request's Body into out.
+func ReadJSON(r *http.Request, out interface{}) error {
+	err := CheckForJSON(r)
+	if err != nil {
+		return err
+	}
+	if r.Body == nil || r.ContentLength == 0 {
 		return nil
 	}
```

Comment:
Right, so this is the tricky one; previously, this check wasn't there (in existing code), and it would either `panic` (if `.Body` was `nil` - although in practice that would not happen), or fall through to the code below, which would mark the request as "invalid" because the body was empty.

There's a test failing because of this change, as it expected it to fail early, but now falls through, producing a different error;

```
=== RUN   TestContainerInvalidJSON//containers/foobar/exec/no_body
=== CONT  TestContainerInvalidJSON//containers/foobar/exec/no_body
    container_test.go:42: assertion failed: string "{\"message\":\"No exec command specified\"}\n" does not contain "invalid JSON: got EOF while reading request body"
        --- FAIL: TestContainerInvalidJSON//containers/foobar/exec/no_body (0.00s)

=== RUN   TestContainerInvalidJSON//exec/foobar/start/no_body
=== CONT  TestContainerInvalidJSON//exec/foobar/start/no_body
    container_test.go:38: assertion failed: 404 (res.StatusCode int) != 400 (http.StatusBadRequest int)
        --- FAIL: TestContainerInvalidJSON//exec/foobar/start/no_body (0.00s)
```

I think _technically_ the new handling is correct; an empty body is not an invalid request (irregardless of the content-type or request type (`POST`, `PUT`, `PATCH`, `DELETE`), but I'll have to update the test to make it pass

Reply:
done 👍